src/fileAttachment.js

```diff
@@ -12,9 +13,16 @@ async function remote_fetch(file) {
 
 async function dsv(file, delimiter, {array = false, typed = false} = {}) {
   const text = await file.text();
-  return (delimiter === "\t"
-      ? (array ? tsvParseRows : tsvParse)
-      : (array ? csvParseRows : csvParse))(text, typed && autoType);
+  const parse = (delimiter === "\t"
+    ? (array ? tsvParseRows : tsvParse)
+    : (array ? csvParseRows : csvParse));
+  if (typed === "auto") {
+    const source = parse(text);
+    const schema = inferSchema(source);
```

Comment:
This is missing one of the things we talked about:

```suggestion
    const schema = inferSchema(source, source.columns);
```

We already know the columns thanks to dsvParse, so we don’t have to scan them again.

Even better, you could change `inferSchema` itself so that it peeks at source.columns rather than always defaulting to getAllKeys in the default expression for `columns`. That might be nice because as it is today, everyone who calls `inferSchema` has to remember to pass in `source.columns` if it exists.

Reply:
Changed inferSchema to default columns to `source.columns || getAllKeys(source)`